File: src/tracer/log.cc
// Copyright 2011 The University of Michigan
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Authors - Jie Yu (jieyu@umich.edu)

// File: tracer/log.cc - Implementation of the trace log.

#include "tracer/log.h"

#include <cassert>
#include <time.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "core/logging.h"

namespace tracer {

#define LOG_SLICE_SIZE  (1024 * 128)

TraceLog::TraceLog(const std::string &path)
    : path_(path),
      mode_(OP_MODE_INVALID),
      meta_(NULL),
      curr_slice_(NULL),
      entry_cursor_(0),
      has_next_(false) {
  // empty
}

void TraceLog::OpenForRead() {
  // set mode
  mode_ = OP_MODE_READ;
  // prepare dir
  PrepareDirForRead();
  // read meta
  std::stringstream meta_ss;
  meta_ss << path_ << "/meta";
  std::fstream meta_in;
  meta_in.open(meta_ss.str().c_str(), std::ios::in | std::ios::binary);
  assert(meta_in.is_open());
  meta_ = new LogMetaProto;
  meta_->ParseFromIstream(&meta_in);
  meta_in.close();
  // read the first slice
  std::stringstream slice_ss;
  slice_ss << path_ << "/1";
  std::fstream slice_in;
  slice_in.open(slice_ss.str().c_str(), std::ios::in | std::ios::binary);
  assert(slice_in.is_open());
  curr_slice_ = new LogSliceProto;
  curr_slice_->ParseFromIstream(&slice_in);
  slice_in.close();
  DEBUG_ASSERT(meta_->uid() == curr_slice_->uid());
  // set cursor
  entry_cursor_ = 0;
  if (curr_slice_->entry_size())
    has_next_ = true;
}

void TraceLog::OpenForWrite() {
  // set mode
  mode_ = OP_MODE_WRITE;
  // clear and create path
  PrepareDirForWrite();
  // create meta
  trace_log_uid_t uid = GenUid();
  meta_ = new LogMetaProto;
  meta_->set_uid(uid);
  meta_->set_slice_count(1);
  // create the current slice
  curr_slice_ = new LogSliceProto;
  curr_slice_->set_uid(uid);
  curr_slice_->set_slice_no(meta_->slice_count());
}

void TraceLog::CloseForRead() {
  // reclaim resource
  curr_slice_->Clear();
  meta_->Clear();
}

void TraceLog::CloseForWrite() {
  // write the current slice
  std::stringstream slice_ss;
  slice_ss << path_ << "/" << std::dec << curr_slice_->slice_no();
  std::fstream slice_out;
  slice_out.open(slice_ss.str().c_str(),
                 std::ios::out | std::ios::trunc | std::ios::binary);
  assert(slice_out.is_open());
  curr_slice_->SerializeToOstream(&slice_out);
  slice_out.close();
  // write meta
  std::stringstream meta_ss;
  meta_ss << path_ << "/meta";
  std::fstream meta_out;
  meta_out.open(meta_ss.str().c_str(),
                std::ios::out | std::ios::trunc | std::ios::binary);
  assert(meta_out.is_open());
  meta_->SerializeToOstream(&meta_out);
  meta_out.close();
  // reclaim resource
  curr_slice_->Clear();
  meta_->Clear();
}

bool TraceLog::HasNextEntry() {
  DEBUG_ASSERT(mode_ == OP_MODE_READ);
  if (!has_next_)
    SwitchSliceForRead();
  return has_next_;
}

LogEntry TraceLog::NextEntry() {
  DEBUG_ASSERT(mode_ == OP_MODE_READ);
  DEBUG_ASSERT(has_next_);
  DEBUG_ASSERT(entry_cursor_ >= 0 && entry_cursor_ < curr_slice_->entry_size());
  LogEntryProto *entry_proto = curr_slice_->mutable_entry(entry_cursor_++);
  if (entry_cursor_ == curr_slice_->entry_size())
    has_next_ = false;
  return LogEntry(entry_proto);
}

LogEntry TraceLog::NewEntry() {
  DEBUG_ASSERT(mode_ == OP_MODE_WRITE);
  if (curr_slice_->entry_size() >= LOG_SLICE_SIZE)
    SwitchSliceForWrite();
  LogEntryProto *entry_proto = curr_slice_->add_entry();
  return LogEntry(entry_proto);
}

trace_log_uid_t TraceLog::GenUid() {
  return (trace_log_uid_t)time(NULL);
}

void TraceLog::SwitchSliceForRead() {
  DEBUG_ASSERT(mode_ == OP_MODE_READ);
  uint32 curr_slice_no = curr_slice_->slice_no();
  uint32 next_slice_no = curr_slice_no + 1;
  curr_slice_->Clear();
  // check whether the slice exists
  std::stringstream slice_ss;
  slice_ss << path_ << "/" << std::dec << next_slice_no;
  std::fstream slice_in;
  slice_in.open(slice_ss.str().c_str(), std::ios::in | std::ios::binary);
  if (slice_in.is_open()) {
    curr_slice_->ParseFromIstream(&slice_in);
    slice_in.close();
    DEBUG_ASSERT(curr_slice_->entry_size());
    entry_cursor_ = 0;
    has_next_ = true;
  } else {
    has_next_ = false;
  }
}

void TraceLog::SwitchSliceForWrite() {
  DEBUG_ASSERT(mode_ == OP_MODE_WRITE);
  uint32 curr_slice_no = curr_slice_->slice_no();
  uint32 next_slice_no = curr_slice_no + 1;
  // save the current slice
  std::stringstream slice_ss;
  slice_ss << path_ << "/" << std::dec << curr_slice_no;
  std::fstream slice_out;
  slice_out.open(slice_ss.str().c_str(),
                 std::ios::out | std::ios::trunc | std::ios::binary);
  assert(slice_out.is_open());
  curr_slice_->SerializeToOstream(&slice_out);
  slice_out.close();
  curr_slice_->Clear();
  // create the new slice
  curr_slice_->set_uid(meta_->uid());
  curr_slice_->set_slice_no(next_slice_no);
  meta_->set_slice_count(next_slice_no);
}

void TraceLog::PrepareDirForRead() {
  DEBUG_ASSERT(mode_ == OP_MODE_READ);
  struct stat sb;
  if (stat(path_.c_str(), &sb) || !S_ISDIR(sb.st_mode)) {
    fprintf(stderr, "please check the trace log dir.\n");
    assert(0);
  }
}

void TraceLog::PrepareDirForWrite() {
  DEBUG_ASSERT(mode_ == OP_MODE_WRITE);
  std::string cmd;
  int res;
  cmd.append("rm -rf ");
  cmd.append(path_);
  res = system(cmd.c_str());
  assert(!res);
  res = mkdir(path_.c_str(), 0755);
  assert(!res);
}

} // namespace tracer

File: script/maple/tracer/log.py
"""Copyright 2011 The University of Michigan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Authors - Jie Yu (jieyu@umich.edu)
"""

import os
from maple.core import proto

def log_pb2():
    return proto.module('tracer.log_pb2')

def log_entry_type_name(t):
    d = log_pb2().LogEntryProto.DESCRIPTOR.fields_by_name['type'].enum_type
    return d.values_by_number[t].name[10:]

class LogEntry(object):
    def __init__(self, proto, log):
        self.proto = proto
        self.log = log
    def type(self):
        return self.proto.type
    def thd_id(self):
        if self.proto.HasField('thd_id'):
            return self.proto.thd_id
        else:
            return None
    def thd_clk(self):
        if self.proto.HasField('thd_clk'):
            return self.proto.thd_clk
        else:
            return None
    def inst(self):
        if self.proto.HasField('inst_id'):
            return self.log.sinfo.find_inst(self.proto.inst_id)
        else:
            return None
    def arg(self, idx):
        return self.proto.arg[idx]
    def num_args(self):
        return len(self.proto.arg)
    def str_arg(self, idx):
        return self.proto.str_arg[idx]
    def num_str_args(self):
        return len(self.proto.str_arg)
    def __str__(self):
        content = []
        content.append('%-27s' % log_entry_type_name(self.type()))
        tid = self.thd_id()
        clk = self.thd_clk()
        ins = self.inst()
        if tid != None:
            content.append('T%x' % tid)
        if clk != None:
            content.append('%08x' % clk)
        if ins != None:
            content.append('[%-40s]' % str(ins))
        for p in self.proto.arg:
            content.append('%x' % p)
        for p in self.proto.str_arg:
            content.append('%s' % p)
        return ' '.join(content)

class TraceLog(object):
    def __init__(self, sinfo):
        self.sinfo = sinfo
        self.meta = log_pb2().LogMetaProto()
        self.slice = log_pb2().LogSliceProto()
        self.mode = None
        self.path = None
        self.entry_cursor = 0
        self.has_next = False
    def open_for_read(self, path):
        if not os.path.isdir(path):
            return False
        meta_path = path + '/meta'
        slice_path = path + '/1'
        if not os.path.exists(meta_path):
            return False
        if not os.path.exists(slice_path):
            return False
        # read meta data
        f = open(meta_path, 'rb')
        self.meta.ParseFromString(f.read())
        f.close()
        # read log slice
        f = open(slice_path, 'rb')
        self.slice.ParseFromString(f.read())
        f.close()
        # setup
        self.mode = 'READ'
        self.path = path
        self.entry_cursor = 0
        if len(self.slice.entry) > 0:
            self.has_next = True
        else:
            self.has_next = False
        return True
    def close_for_read(self):
        assert self.mode == 'READ'
        self.slice.Clear()
        self.meta.Clear()
        self.mode = None
        self.path = None
        self.entry_cursor = 0
        self.has_next = False
    def has_next_entry(self):
        assert self.mode == 'READ'
        if not self.has_next:
            self.switch_slice_for_read()
        return self.has_next
    def next_entry(self):
        assert self.mode == 'READ'
        assert self.has_next
        entry_proto = self.slice.entry[self.entry_cursor]
        entry = LogEntry(entry_proto, self)
        self.entry_cursor += 1
        if self.entry_cursor == len(self.slice.entry):
            self.has_next = False
        return entry
    def switch_slice_for_read(self):
        assert self.mode == 'READ'
        curr_slice_no = self.slice.slice_no
        next_slice_no = curr_slice_no + 1
        self.slice.Clear()
        slice_path = self.path + '/%d' % next_slice_no
        if not os.path.exists(slice_path):
            self.has_next = False
        else:
            f = open(slice_path, 'rb')
            self.slice.ParseFromString(f.read())
            f.close()
            assert len(self.slice.entry) > 0
            self.entry_cursor = 0
            self.has_next = True
    def display(self, f, path):
        self.open_for_read(path)
        while self.has_next_entry():
            entry = self.next_entry()
            f.write('%s\n' % str(entry))
        self.close_for_read()

